usecases/classification/writer.go

```diff
@@ -0,0 +1,170 @@
+//
+
+package classification
+
+import (
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/semi-technologies/weaviate/entities/schema/kind"
+	"github.com/semi-technologies/weaviate/entities/search"
+	"github.com/semi-technologies/weaviate/usecases/kinds"
+)
+
+type writer interface {
+	Start()
+	Store(item search.Result) error
+	Stop() batchWriterResults
+}
+
+type batchWriter struct {
+	mutex           sync.RWMutex
+	vectorRepo      vectorRepo
+	batchItemsCount int
+	batchThings     kinds.BatchThings
+	batchActions    kinds.BatchActions
+	saveThingItems  chan kinds.BatchThings
+	saveActionItems chan kinds.BatchActions
+	errorCount      int64
+	ec              *errorCompounder
+	cancel          chan struct{}
+	batchTreshold   int
+}
+
+type batchWriterResults struct {
+	successCount int64
+	errorCount   int64
+	err          error
+}
+
+func newBatchWriter(vectorRepo vectorRepo) writer {
+	return &batchWriter{
+		vectorRepo:      vectorRepo,
+		batchItemsCount: 0,
+		batchThings:     kinds.BatchThings{},
+		batchActions:    kinds.BatchActions{},
+		saveThingItems:  make(chan kinds.BatchThings, 1),
+		saveActionItems: make(chan kinds.BatchActions, 1),
+		errorCount:      0,
+		ec:              &errorCompounder{},
+		cancel:          make(chan struct{}),
+		batchTreshold:   100,
+	}
+}
+
+// Store puts an item to batch list
+func (r *batchWriter) Store(item search.Result) error {
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+
+	switch item.Kind {
+	case kind.Thing:
+		return r.storeThing(item)
+	case kind.Action:
+		return r.storeAction(item)
+	default:
+		return errors.Errorf("impossible kind")
+	}
+}
+
+// Start starts the batch save goroutine
+func (r *batchWriter) Start() {
+	go r.batchSave()
+}
+
+// Stop stops the batch save goroutine and saves the last items
+func (r *batchWriter) Stop() batchWriterResults {
+	r.cancel <- struct{}{}
+	r.saveThings(r.batchThings)
+	r.saveActions(r.batchActions)
+	return batchWriterResults{int64(r.batchItemsCount) - r.errorCount, r.errorCount, r.ec.toError()}
+}
+
+func (r *batchWriter) storeThing(item search.Result) error {
+	r.batchItemsCount++
+	batchThing := kinds.BatchThing{
+		UUID:          item.ID,
+		Thing:         item.Thing(),
+		OriginalIndex: r.batchItemsCount,
+		Vector:        item.Vector,
+	}
+	r.batchThings = append(r.batchThings, batchThing)
+	if len(r.batchThings) >= r.batchTreshold {
+		r.saveThingItems <- r.batchThings
+		r.batchThings = kinds.BatchThings{}
+	}
+	return nil
+}
+
+func (r *batchWriter) storeAction(item search.Result) error {
+	r.batchItemsCount++
+	batchAction := kinds.BatchAction{
+		UUID:          item.ID,
+		Action:        item.Action(),
+		OriginalIndex: r.batchItemsCount,
+		Vector:        item.Vector,
+	}
+	r.batchActions = append(r.batchActions, batchAction)
+	if len(r.batchActions) >= r.batchTreshold {
+		r.saveActionItems <- r.batchActions
+		r.batchActions = kinds.BatchActions{}
+	}
+	return nil
+}
+
+func (r *batchWriter) batchSave() {
```

Comment:
Very smart to have this non-blocking save logic, but I think we need a comment explaining the motivation behind. Ideally also so that a not-very experienced contributor could also understand it :)

Reply:
Added a comment to that method :)